Generator/VisitFakeQuery.php

```diff
@@ -0,0 +1,357 @@
+            return $this->returnUserIds[array_rand($this->returnUserIds)];
+        }
+
+        $idvisitor = random_bytes(8);
+
+        // 1% chance to update a return user id if the pool is full
+        if (count($this->returnUserIds) >= $this->returnVisitorPoolSize && rand(0,100) === 1) {
+
+            // Never replace the early return visitors, we want some selects at the quiet end of the index
+            $this->returnUserIds[rand($this->returnVisitorPoolSize / 10, $this->returnVisitorPoolSize)] = $idvisitor;
+        }
+
+        // If ppol is not full then add this id
+        if (count($this->returnUserIds) < $this->returnVisitorPoolSize) {
+            $this->returnUserIds[] = $idvisitor;
+        }
+
+        return $idvisitor;
+    }
+
+    public function getCheckActionExistsQuery(string $actionUrl): array
+    {
+
+        $sql = "SELECT `idaction`, `type`, `name` 
+                FROM log_action 
+                WHERE ( hash = CRC32(:name) AND name = :name AND type = 1) OR ( hash = CRC32(:name) AND name = :name AND type = 1)";
+
+        $bind = [
+            ':name' => $actionUrl
+        ];
+        return ['sql' => $sql, 'bind' => $bind];
+    }
+
+    public function getInsertActionQuery(string $actionUrl): array
+    {
+        $sql = "INSERT INTO log_action (name, hash, type, url_prefix) VALUES (:name, CRC32(:name), 1, null);";
+        $bind = [
+            ':name' => $actionUrl
+        ];
+        return ['sql' => $sql, 'bind' => $bind];
+    }
+
+
+    public function getCheckIfNewVisitorQuery(string $idvisitor, int $site): array
+    {
+
+        $timeLookback = date('Y-m-d H:i:s', time() - 1800);
+
+        $sql = "SELECT visit_last_action_time, visit_first_action_time, idvisitor, idvisit, user_id, visit_exit_idaction_url, visit_exit_idaction_name,
+         visitor_returning, visitor_seconds_since_first, visitor_seconds_since_order, visitor_count_visits, visit_goal_buyer, location_country,
+         location_region, location_city, location_latitude, location_longitude, referer_name, referer_keyword, referer_type, idsite, profilable,
+         visit_entry_idaction_url, visit_total_actions, visit_total_interactions, visit_total_searches, referer_url, config_browser_name,
+         config_client_type, config_device_brand, config_device_model, config_device_type, visit_total_events, visit_total_time, location_ip,
+         location_browser_lang, last_idlink_va, custom_dimension_1, custom_dimension_2, custom_dimension_3, custom_dimension_4, custom_dimension_5
+         FROM log_visit 
+         WHERE idsite = :site AND visit_last_action_time >= :lastaction AND idvisitor = UNHEX(:idvisitor) ORDER BY visit_last_action_time DESC LIMIT 1";
+
+        // Removed FORCE INDEX (index_idsite_idvisitor)
+
+        $bind = [':lastaction' => $timeLookback, ':idvisitor' => bin2hex($idvisitor), ':site' => $site];
+        return ['sql' => $sql, 'bind' => $bind];
+    }
+
+    public function getRandomIP() : string
```

Comment:
Could we add support for IPv6? Would it make any difference to anything?

Reply:
I've now added support for IPv6 for a random selection of visits using the same Faker library as the normal visitor generator code :+1: